<h1 align="center">The Battle of Neighborhoods - The Finale</h1>





Author: Abdullah Tasleem

2nd April 2021

# Introduction 

European Countries are quite famous and popular vacation destinations for people all around the world. London, the city of dreams and Paris, the city of love are quite the famous places. They are diverse and multicultural and offer a wide variety of experiences that is widely sought after. We try to group the neighbourhoods of London and Paris respectively and draw insights to what they look like.

# II.	Business Problem

The aim of the project is to help tourists choose their destinations depending on the experiences that the neighbourhoods have to offer and what they would want to have. This also helps people make decisions if they are thinking about migrating to London or Paris or even if they want to relocate neighbourhoods within the city. Relocating requires intensive research for good housing prices and reputated schools for their children. 
Our findings will help stakeholders make informed decisions and address any concerns they have including the different kinds of cuisines, provision stores and what the city has to offer. It will help people to get awareness of the area and neighborhood before moving to a new city, state, country or place for vacation, their work or to start a new fresh life.

# III.	Data

We get geolocation data for London and Paris. Using Postal codes in each city we use can find out the neighbourhoods, boroughs, venues and their most popular venue categories.
For London, we will scrape our data from a Wikipedia page.
For Paris, we will import the dataset available online. 

## Foursquare API
We will need data about different venues in different neighbourhoods of that specific borough. In order to gain that information we will use "Foursquare" locational information. Foursquare is a location data provider with information about all manner of venues and events within an area of interest. Such information includes venue names, locations, menus and even photos. Four-square API as its prime data gathering source as it has a database of millions of places, especially their places API which provides the ability to perform location search, location sharing and details about a business.

We cluster the neighbourhoods together based on similar venue categories. We then present our observations and findings. Using this data, our stakeholders can take the necessary decision.




# Methodology

We will start by importing some of the Python libraries that we need for this project


In [1]:
import pandas as pd
import requests
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means for the clustering stage
from sklearn.cluster import KMeans

In [28]:
!pip install folium
print("Folium installed")
import folium
print("folium is installed")

Folium installed
folium is installed


We will start by exploring the cities individually, then plot the map to show the neighbourhoods, after which we will build our model by clustering the similar neighbourhoods and finally plot a map with the clustered neightbourhoods.

# Neighbourhoods of London

## Data Collection

We start by scraping the wikipedia page on London


In [3]:
#we need Response 200 which means that we are able to make the connection

url_london = "https://en.wikipedia.org/wiki/List_of_areas_of_London"
w_london_url = requests.get(url_london)
w_london_url

<Response [200]>

In [4]:
w_london_data = pd.read_html(w_london_url.text)
w_london_data

[                                                   0
 0  Map all coordinates in "Category:Areas of Lond...
 1                       Download coordinates as: KML,
             Location                     London borough       Post town  \
 0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
 1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
 2          Addington                         Croydon[8]         CROYDON   
 3         Addiscombe                         Croydon[8]         CROYDON   
 4        Albany Park                             Bexley  BEXLEY, SIDCUP   
 ..               ...                                ...             ...   
 526         Woolwich                          Greenwich          LONDON   
 527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
 528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
 529          Yeading                         Hillingdon           HAYES   
 

We nedd the second table on the page to continue, and the below code gets us the second table on the page

In [5]:
w_london_data = w_london_data[1]
w_london_data

Location                     London borough       Post town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS grid ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

## Data Preprocessing

We now work on cleaning the data.
the first thing we do is, remove the spaces between the names of the headers and replace them with "_"

In [6]:
w_london_data.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)
w_london_data

Location                     London borough       Post_town  \
0         Abbey Wood              Bexley, Greenwich [7]          LONDON   
1              Acton  Ealing, Hammersmith and Fulham[8]          LONDON   
2          Addington                         Croydon[8]         CROYDON   
3         Addiscombe                         Croydon[8]         CROYDON   
4        Albany Park                             Bexley  BEXLEY, SIDCUP   
..               ...                                ...             ...   
526         Woolwich                          Greenwich          LONDON   
527   Worcester Park       Sutton, Kingston upon Thames  WORCESTER PARK   
528  Wormwood Scrubs             Hammersmith and Fulham          LONDON   
529          Yeading                         Hillingdon           HAYES   
530         Yiewsley                         Hillingdon    WEST DRAYTON   

    Postcode district Dial code OS_grid_ref  
0                 SE2       020    TQ465785  
1              W3, W4       020    TQ205805  
2                 CR0       020    TQ375645  
3                 CR0       020    TQ345665  
4           DA5, DA14       020    TQ478728  
..                ...       ...         ...  
526              SE18       020    TQ435795  
527               KT4       020    TQ225655  
528               W12       020    TQ225815  
529               UB4       020    TQ115825  
530               UB7       020    TQ063804  

[531 rows x 6 columns]

Notice that a few columns still have spaces. 
This means there are special characters.

## Feature Selection

We only need Boroughs, Postal codes and post town columns, so we will keep them and drop the rest.

In [7]:
df1 = w_london_data.drop(
    [ w_london_data.columns[0], w_london_data.columns[4], w_london_data.columns[5] ], axis=1)

df1.head()


London borough       Post_town Postcode district
0              Bexley, Greenwich [7]          LONDON               SE2
1  Ealing, Hammersmith and Fulham[8]          LONDON            W3, W4
2                         Croydon[8]         CROYDON               CR0
3                         Croydon[8]         CROYDON               CR0
4                             Bexley  BEXLEY, SIDCUP         DA5, DA14

We rename the columns to remove the spaces in between and have simpler names.

In [8]:
df1.columns = ['borough','town','post_code']
df1

borough            town  post_code
0                Bexley, Greenwich [7]          LONDON        SE2
1    Ealing, Hammersmith and Fulham[8]          LONDON     W3, W4
2                           Croydon[8]         CROYDON        CR0
3                           Croydon[8]         CROYDON        CR0
4                               Bexley  BEXLEY, SIDCUP  DA5, DA14
..                                 ...             ...        ...
526                          Greenwich          LONDON       SE18
527       Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528             Hammersmith and Fulham          LONDON        W12
529                         Hillingdon           HAYES        UB4
530                         Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

We need to strip down the borough column to get rid of the [] and numbers in them.

In [9]:
df1['borough'] = df1['borough'].map(lambda x: x.rstrip(']').rstrip('0123456789').rstrip('['))
df1

borough            town  post_code
0                Bexley, Greenwich           LONDON        SE2
1    Ealing, Hammersmith and Fulham          LONDON     W3, W4
2                           Croydon         CROYDON        CR0
3                           Croydon         CROYDON        CR0
4                            Bexley  BEXLEY, SIDCUP  DA5, DA14
..                              ...             ...        ...
526                       Greenwich          LONDON       SE18
527    Sutton, Kingston upon Thames  WORCESTER PARK        KT4
528          Hammersmith and Fulham          LONDON        W12
529                      Hillingdon           HAYES        UB4
530                      Hillingdon    WEST DRAYTON        UB7

[531 rows x 3 columns]

In [10]:
# We will get the dimensions of the dataframe

df1.shape

(531, 3)

We now have 531 records and 3 columns.

## Feature Engineering

We only need the neighbourhoods of London so we only get the records which contains London in the town column


In [11]:
df1 = df1[df1['town'].str.contains('LONDON')]
df1

borough                    town post_code
0                Bexley, Greenwich                   LONDON       SE2
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4
6                              City                  LONDON       EC3
7                       Westminster                  LONDON       WC2
9                           Bromley                  LONDON      SE20
..                              ...                     ...       ...
521                       Redbridge                  LONDON  IG8, E18
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8
525                          Barnet                  LONDON       N12
526                       Greenwich                  LONDON      SE18
528          Hammersmith and Fulham                  LONDON       W12

[308 rows x 3 columns]

In [12]:
df1.shape

(308, 3)

We now have 308 records, and all these are the neighbours of London. 
We will now get some descriptive statistics.

In [13]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 308 entries, 0 to 528
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   borough    308 non-null    object
 1   town       308 non-null    object
 2   post_code  308 non-null    object
dtypes: object(3)
memory usage: 9.6+ KB


# Geolocations of the neighbourhoods of London

## ArcGis API

We need to get the geographical co-ordinates for the neighbourhoods to plot out map. We will use the arcgis package to do so.

In [14]:
pip install arcgis

Note: you may need to restart the kernel to use updated packages.


In [15]:
from arcgis.geocoding import geocode
from arcgis.gis import GIS
gis = GIS()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/arcgis/features/_data/geodataset/utils.py:16: FutureWarning: The pandas.datetime class is deprecated and will be removed from pandas in a future version. Import from datetime module instead.
  pd.datetime,


We need the latitude and longitude so we define the arcgis geocode function

In [16]:
def get_x_y_uk(address1):
   lat_coords = 0
   lng_coords = 0
   g = geocode(address='{}, London, England, GBR'.format(address1))[0]
   lng_coords = g['location']['x']
   lat_coords = g['location']['y']
   return str(lat_coords) +","+ str(lng_coords)

In [17]:
c = get_x_y_uk('SE2')
c

'51.492450000000076,0.12127000000003818'

We need to pass the postal codes to the function so we copy the codes

In [18]:
geo_coordinates_uk = df1['post_code']    
geo_coordinates_uk

0           SE2
1        W3, W4
6           EC3
7           WC2
9          SE20
         ...   
521    IG8, E18
522         IG8
525         N12
526        SE18
528         W12
Name: post_code, Length: 308, dtype: object

In [19]:
coordinates_latlng_uk = geo_coordinates_uk.apply(lambda x: get_x_y_uk(x))
coordinates_latlng_uk

0       51.492450000000076,0.12127000000003818
1        51.51324000000005,-0.2674599999999714
6       51.51200000000006,-0.08057999999994081
7       51.51651000000004,-0.11967999999995982
9       51.41009000000008,-0.05682999999993399
                        ...                   
521    51.589770000000044,0.030520000000024083
522      51.50642000000005,-0.1272099999999341
525     51.615920000000074,-0.1767399999999384
526      51.48207000000008,0.07143000000002075
528      51.50645000000003,-0.2369099999999662
Name: post_code, Length: 308, dtype: object

Next we extract the Longitude and Latitude individually.

Once we get the geographical co-ordinates of the neighbourhoods, we then merge our source data and geographical coordinates.

In [20]:
lat_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[0])
lat_uk

0      51.492450000000076
1       51.51324000000005
6       51.51200000000006
7       51.51651000000004
9       51.41009000000008
              ...        
521    51.589770000000044
522     51.50642000000005
525    51.615920000000074
526     51.48207000000008
528     51.50645000000003
Name: post_code, Length: 308, dtype: object

In [21]:
lng_uk = coordinates_latlng_uk.apply(lambda x: x.split(',')[1])
lng_uk

0       0.12127000000003818
1       -0.2674599999999714
6      -0.08057999999994081
7      -0.11967999999995982
9      -0.05682999999993399
               ...         
521    0.030520000000024083
522     -0.1272099999999341
525     -0.1767399999999384
526     0.07143000000002075
528     -0.2369099999999662
Name: post_code, Length: 308, dtype: object

In [22]:
london_merged = pd.concat([df1,lat_uk.astype(float), lng_uk.astype(float)], axis=1)
london_merged.columns= ['borough','town','post_code','latitude','longitude']
london_merged

borough                    town post_code  \
0                Bexley, Greenwich                   LONDON       SE2   
1    Ealing, Hammersmith and Fulham                  LONDON    W3, W4   
6                              City                  LONDON       EC3   
7                       Westminster                  LONDON       WC2   
9                           Bromley                  LONDON      SE20   
..                              ...                     ...       ...   
521                       Redbridge                  LONDON  IG8, E18   
522       Redbridge, Waltham Forest  LONDON, WOODFORD GREEN       IG8   
525                          Barnet                  LONDON       N12   
526                       Greenwich                  LONDON      SE18   
528          Hammersmith and Fulham                  LONDON       W12   

     latitude  longitude  
0    51.49245    0.12127  
1    51.51324   -0.26746  
6    51.51200   -0.08058  
7    51.51651   -0.11968  
9    51.41009   -0.05683  
..        ...        ...  
521  51.58977    0.03052  
522  51.50642   -0.12721  
525  51.61592   -0.17674  
526  51.48207    0.07143  
528  51.50645   -0.23691  

[308 rows x 5 columns]

Getting the geocode for London to help visualize it on map.

In [23]:
london = geocode(address='London, England, GBR')[0]
london_lng_coords = london['location']['x']
london_lat_coords = london['location']['y']
london_lng_coords

-0.1272099999999341

In [24]:
london_lat_coords

51.50642000000005

# Visualization

Folium helps us visualize the map of London and it's neighbourhoods.

In [29]:
# Creating the map of London
map_London = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)
map_London

# adding markers to map
for latitude, longitude, borough, town in zip(london_merged['latitude'], london_merged['longitude'], london_merged['borough'], london_merged['town']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_London)  
    
map_London

# Using Foursquare API


Using Foursquare API we will get the venues and the categories of each neighbourhood in London.

In [30]:
CLIENT_ID = 'AGHGCCBFEW1K5KOBLVNNDLME3WJCBJBEEH4D2L24Y2GNCHXF' 
CLIENT_SECRET = 'X3TCFLP35J0C3ZZXXVYM2FP5OWXB1HDLRXPWS31UASMTAEM2'
VERSION = '20180605' # Foursquare API version

Using the below function we will get the venue categories in the neighbourhood

In [31]:
LIMIT=100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius,
            LIMIT
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

In [32]:
# Getting venues in London

venues_in_London = getNearbyVenues(london_merged['borough'], london_merged['latitude'], london_merged['longitude'])

Bexley, Greenwich 
Ealing, Hammersmith and Fulham
City
Westminster
Bromley
Islington
Islington
Barnet
Enfield
Wandsworth
Southwark
City
Richmond upon Thames
Barnet
Islington
Wandsworth
Westminster
Bromley
Newham
Ealing
Westminster
Lewisham
Camden
Southwark
Tower Hamlets
Bexley
City
Lewisham
Greenwich
Tower Hamlets
Camden
Haringey
Tower Hamlets
Haringey
Barnet
Brent
Lambeth
Lewisham
Tower Hamlets
Kensington and Chelsea, Hammersmith and Fulham
Brent
Barnet
Barnet
Southwark
Tower Hamlets
Camden
Tower Hamlets
Waltham Forest
Newham
Islington
Richmond upon Thames
Lewisham
Camden
Westminster
Greenwich
Kensington and Chelsea
Barnet
Westminster
Lewisham
Waltham Forest
Hounslow, Ealing, Hammersmith and Fulham
Brent
Barnet
Lambeth, Wandsworth
Islington
Barnet
Merton
Barnet
Westminster
Barnet, Brent, Camden
Lewisham
Bexley
Haringey
Bromley
Tower Hamlets
Newham
Hackney
Islington
Southwark
Lewisham
Brent
Southwark
Ealing
Kensington and Chelsea
Wandsworth
Southwark
Barnet
Newham
Richmond upon Thames


In [33]:
venues_in_London.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Bexley, Greenwich                 51.49245                  0.12127   
1  Bexley, Greenwich                 51.49245                  0.12127   
2  Bexley, Greenwich                 51.49245                  0.12127   
3  Bexley, Greenwich                 51.49245                  0.12127   
4  Bexley, Greenwich                 51.49245                  0.12127   

                              Venue     Venue Category  
0                      Lesnes Abbey      Historic Site  
1                       Sainsbury's        Supermarket  
2                              Lidl        Supermarket  
3  Abbey Wood Railway Station (ABW)      Train Station  
4                        Costcutter  Convenience Store

In [34]:
venues_in_London.shape

(10352, 5)

We have 10352 venue records.

# Grouping the Venues by Categories

This will give us some information about the number of categories

In [35]:
venues_in_London.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                        
Accessories Store               Westminster                51.51656   
Adult Boutique                    Islington                51.52969   
African Restaurant              Westminster                51.52587   
American Restaurant          Waltham Forest                51.61780   
Antique Shop                    Westminster                51.55506   
...                                     ...                     ...   
Wings Joint          Hammersmith and Fulham                51.54187   
Women's Store                   Westminster                51.55457   
Xinjiang Restaurant               Southwark                51.47480   
Yoga Studio                     Westminster                51.55457   
Zoo Exhibit                          Camden                51.53354   

                     Neighbourhood Longitude                        Venue  
Venue Category                                                             
Accessories Store                   -0.14770                   Balenciaga  
Adult Boutique                      -0.08697  Sh! Women's Erotic Emporium  
African Restaurant                  -0.08808           Red Sea Restaurant  
American Restaurant                  0.02795                  Spielburger  
Antique Shop                        -0.11968     The London Silver Vaults  
...                                      ...                          ...  
Wings Joint                         -0.19795                     Wingmans  
Women's Store                        0.00278           Vivien of Holloway  
Xinjiang Restaurant                 -0.09313                    Silk Road  
Yoga Studio                         -0.03558                    yogahaven  
Zoo Exhibit                         -0.14606                Penguin Beach  

[300 rows x 4 columns]

We have 300 records.

# One Hot Encoding

This will provide better results for clustering

In [36]:
London_venue_cat = pd.get_dummies(venues_in_London[['Venue Category']], prefix="", prefix_sep="")
London_venue_cat

Accessories Store  Adult Boutique  African Restaurant  \
0                      0               0                   0   
1                      0               0                   0   
2                      0               0                   0   
3                      0               0                   0   
4                      0               0                   0   
...                  ...             ...                 ...   
10347                  0               0                   0   
10348                  0               0                   0   
10349                  0               0                   0   
10350                  0               0                   0   
10351                  0               0                   0   

       American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                        0             0       0                 0   
1                        0             0       0                 0   
2                        0             0       0                 0   
3                        0             0       0                 0   
4                        0             0       0                 0   
...                    ...           ...     ...               ...   
10347                    0             0       0                 0   
10348                    0             0       0                 0   
10349                    0             0       0                 0   
10350                    0             0       0                 0   
10351                    0             0       0                 0   

       Argentinian Restaurant  Art Gallery  Art Museum  ...  \
0                           0            0           0  ...   
1                           0            0           0  ...   
2                           0            0           0  ...   
3                           0            0           0  ...   
4                           0            0           0  ...   
...                       ...          ...         ...  ...   
10347                       0            0           0  ...   
10348                       0            0           0  ...   
10349                       0            0           0  ...   
10350                       0            0           0  ...   
10351                       0            0           0  ...   

       Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  \
0                          0                0           0         0   
1                          0                0           0         0   
2                          0                0           0         0   
3                          0                0           0         0   
4                          0                0           0         0   
...                      ...              ...         ...       ...   
10347                      0                0           0         0   
10348                      0                0           0         0   
10349                      0                0           0         0   
10350                      0                0           0         0   
10351                      0                0           0         0   

       Wine Shop  Wings Joint  Women's Store  Xinjiang Restaurant  \
0              0            0              0                    0   
1              0            0              0                    0   
2              0            0              0                    0   
3              0            0              0                    0   
4              0            0              0                    0   
...          ...          ...            ...                  ...   
10347          0            0              0                    0   
10348          0            0              0                    0   
10349          0            0              0                    0   
10350          0            0              0                    0   
10351          0            

In [37]:
London_venue_cat['Neighbourhood'] = venues_in_London['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [London_venue_cat.columns[-1]] + list(London_venue_cat.columns[:-1])
London_venue_cat = London_venue_cat[fixed_columns]

London_venue_cat.head()

Neighbourhood  Accessories Store  Adult Boutique  African Restaurant  \
0  Bexley, Greenwich                   0               0                   0   
1  Bexley, Greenwich                   0               0                   0   
2  Bexley, Greenwich                   0               0                   0   
3  Bexley, Greenwich                   0               0                   0   
4  Bexley, Greenwich                   0               0                   0   

   American Restaurant  Antique Shop  Arcade  Arepa Restaurant  \
0                    0             0       0                 0   
1                    0             0       0                 0   
2                    0             0       0                 0   
3                    0             0       0                 0   
4                    0             0       0                 0   

   Argentinian Restaurant  Art Gallery  ...  Vietnamese Restaurant  \
0                       0            0  ...                      0   
1                       0            0  ...                      0   
2                       0            0  ...                      0   
3                       0            0  ...                      0   
4                       0            0  ...                      0   

   Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  Wings Joint  \
0                0           0         0          0            0   
1                0           0         0          0            0   
2                0           0         0          0            0   
3                0           0         0          0            0   
4                0           0         0          0            0   

   Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0              0                    0            0            0  
1              0                    0            0            0  
2              0                    0            0            0  
3              0                    0            0            0  
4              0                    0            0            0  

[5 rows x 301 columns]

In [44]:
London_grouped = London_venue_cat.groupby('Neighbourhood').mean().reset_index()
London_grouped.head()

Neighbourhood  Accessories Store  Adult Boutique  \
0                 Barnet                0.0             0.0   
1  Barnet, Brent, Camden                0.0             0.0   
2                 Bexley                0.0             0.0   
3      Bexley, Greenwich                0.0             0.0   
4     Bexley, Greenwich                 0.0             0.0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                 0.0             0.001812           0.0     0.0   
1                 0.0             0.000000           0.0     0.0   
2                 0.0             0.000000           0.0     0.0   
3                 0.0             0.000000           0.0     0.0   
4                 0.0             0.000000           0.0     0.0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  ...  \
0               0.0                0.007246          0.0  ...   
1               0.0                0.000000          0.0  ...   
2               0.0                0.000000          0.0  ...   
3               0.0                0.000000          0.0  ...   
4               0.0                0.000000          0.0  ...   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0               0.001812              0.0         0.0       0.0        0.0   
1               0.000000              0.0         0.0       0.0        0.0   
2               0.000000              0.0         0.0       0.0        0.0   
3               0.000000              0.0         0.0       0.0        0.0   
4               0.000000              0.0         0.0       0.0        0.0   

   Wings Joint  Women's Store  Xinjiang Restaurant  Yoga Studio  Zoo Exhibit  
0          0.0            0.0                  0.0          0.0          0.0  
1          0.0            0.0                  0.0          0.0          0.0  
2          0.0            0.0                  0.0          0.0          0.0  
3          0.0            0.0                  0.0          0.0          0.0  
4          0.0            0.0                  0.0          0.0          0.0  

[5 rows x 301 columns]

In [39]:
#We will define a function to get the top most common venue categories

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

We will take the top 10 venue categories to cluster the neighbourhood. 
We will define a function to label the columns of venue

In [43]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

In [45]:
#Getting the top venue categories in London

# create a new dataframe for London
neighborhoods_venues_sorted_london = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_london['Neighbourhood'] = London_grouped['Neighbourhood']

for ind in np.arange(London_grouped.shape[0]):
    neighborhoods_venues_sorted_london.iloc[ind, 1:] = return_most_common_venues(London_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_london.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0                 Barnet           Coffee Shop                  Café   
1  Barnet, Brent, Camden  Gym / Fitness Center                Bakery   
2                 Bexley           Supermarket         Historic Site   
3      Bexley, Greenwich              Bus Stop                  Park   
4     Bexley, Greenwich            Supermarket         Historic Site   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Grocery Store              Bus Stop                   Pub   
1        Hardware Store        Clothing Store           Supermarket   
2     Convenience Store         Train Station              Platform   
3               Daycare        Massage Studio           Sports Club   
4     Convenience Store              Platform         Train Station   

  6th Most Common Venue       7th Most Common Venue 8th Most Common Venue  \
0           Supermarket          Italian Restaurant              Pharmacy   
1     Fish & Chips Shop          Falafel Restaurant        Farmers Market   
2           Golf Course  Construction & Landscaping                  Park   
3           Golf Course  Construction & Landscaping         Historic Site   
4               Exhibit          Falafel Restaurant        Farmers Market   

  9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant     Turkish Restaurant  
1  Fast Food Restaurant    Filipino Restaurant  
2              Bus Stop      Food & Drink Shop  
3     Fish & Chips Shop         Farmers Market  
4  Fast Food Restaurant    Filipino Restaurant

# Model Building

## K Means

keeping n as 5

In [46]:
# set number of clusters
k_num_clusters = 5

London_grouped_clustering = London_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_london = KMeans(n_clusters=k_num_clusters, random_state=0).fit(London_grouped_clustering)
kmeans_london

KMeans(n_clusters=5, random_state=0)

In [47]:
# Labelling clustered data
kmeans_london.labels_

array([4, 1, 2, 4, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 3, 4, 4, 4, 4, 4,
       4, 4, 4, 0, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4,
       4, 4, 4, 4, 4, 4], dtype=int32)

In [48]:
neighborhoods_venues_sorted_london.insert(0, 'Cluster Labels', kmeans_london.labels_ +1)

Join London_merged with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [49]:
london_data = london_merged

london_data = london_data.join(neighborhoods_venues_sorted_london.set_index('Neighbourhood'), on='borough')

london_data.head()

borough    town post_code  latitude  longitude  \
0              Bexley, Greenwich   LONDON       SE2  51.49245    0.12127   
1  Ealing, Hammersmith and Fulham  LONDON    W3, W4  51.51324   -0.26746   
6                            City  LONDON       EC3  51.51200   -0.08058   
7                     Westminster  LONDON       WC2  51.51651   -0.11968   
9                         Bromley  LONDON      SE20  51.41009   -0.05683   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               3           Supermarket         Historic Site   
1               4         Grocery Store        Breakfast Spot   
6               5           Coffee Shop                 Hotel   
7               5           Coffee Shop                 Hotel   
9               5           Supermarket                 Hotel   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Convenience Store              Platform         Train Station   
1         Train Station                  Park     Indian Restaurant   
6    Italian Restaurant  Gym / Fitness Center                   Pub   
7        Sandwich Place                   Pub                  Café   
9  Fast Food Restaurant         Grocery Store     Convenience Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0               Exhibit    Falafel Restaurant        Farmers Market   
1           Film Studio               Exhibit    Falafel Restaurant   
6        Sandwich Place            Restaurant              Wine Bar   
7    Italian Restaurant               Theater             Hotel Bar   
9                  Park    Italian Restaurant             Pet Store   

  9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant    Filipino Restaurant  
1        Farmers Market               Fountain  
6     French Restaurant                 Garden  
7            Restaurant           Burger Joint  
9                  Café      Fish & Chips Shop

In [50]:
#Drop NaN values 

london_data_nonan = london_data.dropna(subset=['Cluster Labels'])

## Visualize the cluster

In [51]:
map_clusters_london = folium.Map(location=[london_lat_coords, london_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(london_data_nonan['latitude'], london_data_nonan['longitude'], london_data_nonan['borough'], london_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters_london)
        
map_clusters_london

## Examine the clusters

Cluster 1, 3 and 5

### Cluster 1

In [52]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 1, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
377  HARROW, STANMOREEDGWARE, LONDON               1                Bakery   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
377         Metro Station                   Gym              Fountain   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
377            Food Truck            Food Stand            Food Court   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
377     Food & Drink Shop           Flower Shop                Exhibit

### Cluster 3


In [53]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 3, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
0                 LONDON               3           Supermarket   
45   BEXLEYHEATH, LONDON               3           Supermarket   
124               LONDON               3           Supermarket   
291       LONDON, SIDCUP               3           Supermarket   
505               LONDON               3           Supermarket   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Historic Site     Convenience Store              Platform   
45          Historic Site     Convenience Store         Train Station   
124         Historic Site     Convenience Store         Train Station   
291         Historic Site     Convenience Store         Train Station   
505         Historic Site     Convenience Store         Train Station   

    5th Most Common Venue 6th Most Common Venue       7th Most Common Venue  \
0           Train Station               Exhibit          Falafel Restaurant   
45               Platform           Golf Course  Construction & Landscaping   
124              Platform           Golf Course  Construction & Landscaping   
291              Platform           Golf Course  Construction & Landscaping   
505              Platform           Golf Course  Construction & Landscaping   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0          Farmers Market  Fast Food Restaurant    Filipino Restaurant  
45                   Park              Bus Stop      Food & Drink Shop  
124                  Park              Bus Stop      Food & Drink Shop  
291                  Park              Bus Stop      Food & Drink Shop  
505                  Park              Bus Stop      Food & Drink Shop

### Cluster 5


In [54]:
london_data_nonan.loc[london_data_nonan['Cluster Labels'] == 5, london_data_nonan.columns[[1] + list(range(5, london_data_nonan.shape[1]))]]

town  Cluster Labels 1st Most Common Venue  \
6                    LONDON               5           Coffee Shop   
7                    LONDON               5           Coffee Shop   
9                    LONDON               5           Supermarket   
10                   LONDON               5           Coffee Shop   
12                   LONDON               5           Coffee Shop   
..                      ...             ...                   ...   
521                  LONDON               5                   Pub   
522  LONDON, WOODFORD GREEN               5                 Hotel   
525                  LONDON               5           Coffee Shop   
526                  LONDON               5                   Pub   
528                  LONDON               5                   Pub   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Hotel    Italian Restaurant  Gym / Fitness Center   
7                   Hotel        Sandwich Place                   Pub   
9                   Hotel  Fast Food Restaurant         Grocery Store   
10                    Pub                  Café            Food Truck   
12                    Pub                  Café            Food Truck   
..                    ...                   ...                   ...   
521         Grocery Store                Bakery                  Café   
522   Monument / Landmark                Garden               Theater   
525                  Café         Grocery Store              Bus Stop   
526         Grocery Store              Bus Stop     Indian Restaurant   
528                  Café           Coffee Shop         Grocery Store   

     5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6                      Pub        Sandwich Place            Restaurant   
7                     Café    Italian Restaurant               Theater   
9        Convenience Store                  Park    Italian Restaurant   
10   Vietnamese Restaurant                  Park    Italian Restaurant   
12   Vietnamese Restaurant                  Park    Italian Restaurant   
..                     ...                   ...                   ...   
521            Pizza Place    Seafood Restaurant             BBQ Joint   
522                    Pub                 Plaza                  Café   
525                    Pub           Supermarket    Italian Restaurant   
526            Coffee Shop                 Hotel         Historic Site   
528              Gastropub       Thai Restaurant                 Hotel   

    8th Most Common Venue 9th Most Common Venue      10th Most Common Venue  
6                Wine Bar     French Restaurant                      Garden  
7               Hotel Bar            Restaurant                Burger Joint  
9               Pet Store                  Café           Fish & Chips Shop  
10           Cocktail Bar                 Hotel        Gym / Fitness Center  
12           Cocktail Bar                 Hotel        Gym / Fitness Center  
..                    ...                   ...                         ...  
521         Metro Station           Bridal Shop                         Bar  
522        Sandwich Place                Bakery                 Coffee Shop  
525              Pharmacy      Sushi Restaurant          Turkish Restaurant  
526           Auto Garage    Turkish Restaurant  Construction & Landscaping  
528                Bakery    Italian Restaurant                 Pizza Place  

[300 rows x 12 columns]

# Paris

## Neighbourhoods or Paris

### Data Collection

We use pandas to read the json file

In [56]:
!wget -q -O 'france-data.json' https://www.data.gouv.fr/fr/datasets/r/e88c6fda-1d09-42a0-a069-606d3259114e
print("Data Downloaded!")


Data Downloaded!


In [57]:
paris_raw = pd.read_json('france-data.json')
paris_raw.head()

datasetid  \
0  correspondances-code-insee-code-postal   
1  correspondances-code-insee-code-postal   
2  correspondances-code-insee-code-postal   
3  correspondances-code-insee-code-postal   
4  correspondances-code-insee-code-postal   

                                   recordid  \
0  2bf36b38314b6c39dfbcd09225f97fa532b1fc45   
1  7ee82e74e059b443df18bb79fc5a19b1f05e5a88   
2  e2cd3186f07286705ed482a10b6aebd9de633c81   
3  868bf03527a1d0a9defe5cf4e6fa0a730d725699   
4  21e809b1d4480333c8b6fe7addd8f3b06f343e2c   

                                              fields  \
0  {'code_comm': '645', 'nom_dept': 'ESSONNE', 's...   
1  {'code_comm': '133', 'nom_dept': 'SEINE-ET-MAR...   
2  {'code_comm': '378', 'nom_dept': 'ESSONNE', 's...   
3  {'code_comm': '243', 'nom_dept': 'SEINE-ET-MAR...   
4  {'code_comm': '003', 'nom_dept': 'VAL-DE-MARNE...   

                                            geometry  \
0  {'type': 'Point', 'coordinates': [2.2517129721...   
1  {'type': 'Point', 'coordinates': [3.0529405055...   
2  {'type': 'Point', 'coordinates': [2.1971816504...   
3  {'type': 'Point', 'coordinates': [2.7097808131...   
4  {'type': 'Point', 'coordinates': [2.3335102498...   

                record_timestamp  
0  2016-09-21T00:29:06.175+02:00  
1  2016-09-21T00:29:06.175+02:00  
2  2016-09-21T00:29:06.175+02:00  
3  2016-09-21T00:29:06.175+02:00  
4  2016-09-21T00:29:06.175+02:00

## Data preprocessing

We need to break down each field to create the dataframe for use

In [58]:
paris_field_data = pd.DataFrame()
for f in paris_raw.fields:
    dict_new = f
    paris_field_data = paris_field_data.append(dict_new, ignore_index=True)

paris_field_data.head()

code_arr code_cant code_comm code_dept code_reg  \
0        3        03       645        91       11   
1        3        20       133        77       11   
2        1        09       378        91       11   
3        5        14       243        77       11   
4        3        34       003        94       11   

                              geo_point_2d  \
0  [48.750443119964764, 2.251712972144151]   
1   [48.41256065214989, 3.052940505560729]   
2    [48.52726809075556, 2.19718165044305]   
3  [48.87307018579678, 2.7097808131278462]   
4   [48.80588035965699, 2.333510249842654]   

                                           geo_shape id_geofla insee_com  \
0  {'type': 'Polygon', 'coordinates': [[[2.238024...     16275     91645   
1  {'type': 'Polygon', 'coordinates': [[[3.076046...     31428     77133   
2  {'type': 'Polygon', 'coordinates': [[[2.203466...     30975     91378   
3  {'type': 'Polygon', 'coordinates': [[[2.727542...     17000     77243   
4  {'type': 'Polygon', 'coordinates': [[[2.343851...     32123     94003   

               nom_comm        nom_dept     nom_region  population  \
0  VERRIERES-LE-BUISSON         ESSONNE  ILE-DE-FRANCE        15.5   
1  COURCELLES-EN-BASSEE  SEINE-ET-MARNE  ILE-DE-FRANCE         0.2   
2             MAUCHAMPS         ESSONNE  ILE-DE-FRANCE         0.3   
3       LAGNY-SUR-MARNE  SEINE-ET-MARNE  ILE-DE-FRANCE        20.2   
4               ARCUEIL    VAL-DE-MARNE  ILE-DE-FRANCE        19.5   

  postal_code            statut  superficie  z_moyen  
0       91370    Commune simple       999.0    121.0  
1       77126    Commune simple      1082.0     88.0  
2       91730    Commune simple       313.0    150.0  
3       77400  Chef-lieu canton       579.0     71.0  
4       94110  Chef-lieu canton       232.0     70.0

## Feature Selection

The columns needed for this section are geo_point_2nd, nom_dept, nom_comm and postal_code

In [59]:
df_2 = paris_field_data[['postal_code','nom_comm','nom_dept','geo_point_2d']]
df_2

postal_code              nom_comm           nom_dept  \
0          91370  VERRIERES-LE-BUISSON            ESSONNE   
1          77126  COURCELLES-EN-BASSEE     SEINE-ET-MARNE   
2          91730             MAUCHAMPS            ESSONNE   
3          77400       LAGNY-SUR-MARNE     SEINE-ET-MARNE   
4          94110               ARCUEIL       VAL-DE-MARNE   
...          ...                   ...                ...   
1295       77520     CESSOY-EN-MONTOIS     SEINE-ET-MARNE   
1296       93420            VILLEPINTE  SEINE-SAINT-DENIS   
1297       77130         CANNES-ECLUSE     SEINE-ET-MARNE   
1298       78930              VILLETTE           YVELINES   
1299       95270  LE PLESSIS-LUZARCHES         VAL-D'OISE   

                                 geo_point_2d  
0     [48.750443119964764, 2.251712972144151]  
1      [48.41256065214989, 3.052940505560729]  
2       [48.52726809075556, 2.19718165044305]  
3     [48.87307018579678, 2.7097808131278462]  
4      [48.80588035965699, 2.333510249842654]  
...                                       ...  
1295   [48.50730730461658, 3.138844194183689]  
1296   [48.95902025378707, 2.536306342059409]  
1297   [48.36403767307805, 2.990786679832767]  
1298  [48.92627887061508, 1.6937417245662671]  
1299  [49.09572967201378, 2.4547564431234923]  

[1300 rows x 4 columns]

We have managed to collect 1300 records.

## Feature Engineering


We localize to only include Paris

In [61]:
df_paris = df_2[df_2['nom_dept'].str.contains('PARIS')].reset_index(drop=True)
df_paris

postal_code                  nom_comm nom_dept  \
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS   
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS   
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS   
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS   
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS   
5        75004   PARIS-4E-ARRONDISSEMENT    PARIS   
6        75008   PARIS-8E-ARRONDISSEMENT    PARIS   
7        75013  PARIS-13E-ARRONDISSEMENT    PARIS   
8        75012  PARIS-12E-ARRONDISSEMENT    PARIS   
9        75005   PARIS-5E-ARRONDISSEMENT    PARIS   
10       75019  PARIS-19E-ARRONDISSEMENT    PARIS   
11       75020  PARIS-20E-ARRONDISSEMENT    PARIS   
12       75010  PARIS-10E-ARRONDISSEMENT    PARIS   
13       75016  PARIS-16E-ARRONDISSEMENT    PARIS   
14       75007   PARIS-7E-ARRONDISSEMENT    PARIS   
15       75018  PARIS-18E-ARRONDISSEMENT    PARIS   
16       75017  PARIS-17E-ARRONDISSEMENT    PARIS   
17       75015  PARIS-15E-ARRONDISSEMENT    PARIS   
18       75001  PARIS-1ER-ARRONDISSEMENT    PARIS   
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS   

                               geo_point_2d  
0    [48.87689616237872, 2.337460241388529]  
1    [48.86790337886785, 2.344107166658533]  
2    [48.85941549762748, 2.378741060237548]  
3    [48.86305413181178, 2.359361058970589]  
4    [48.84896809191946, 2.332670898588416]  
5   [48.854228281954754, 2.357361938142205]  
6    [48.87252726662346, 2.312582560420059]  
7    [48.82871768452136, 2.362468228516128]  
8    [48.83515623066034, 2.419807034965275]  
9   [48.844508659617546, 2.349859385560182]  
10   [48.88686862295828, 2.384694327870042]  
11   [48.86318677744551, 2.400819826729021]  
12   [48.87602855694339, 2.361112904561707]  
13   [48.86039876035177, 2.262099559395783]  
14   [48.85608259819694, 2.312438687733857]  
15  [48.892735074561706, 2.348711933867703]  
16   [48.88733716648682, 2.307485559493426]  
17   [48.84015541860987, 2.293559372435076]  
18    [48.8626304851685, 2.336293446550539]  
19   [48.82899321160942, 2.327100883257538]

In [62]:
df_paris.shape

(20, 4)

We have managed to bring down the records to just 20 from 1300. Paris data is quite small than initally thought

## Gelocations of the Neighbourhoods of Paris

Checking one of the geo coordinates.

In [63]:
df_paris['geo_point_2d'][0]

[48.87689616237872, 2.337460241388529]

In [64]:
temp1 = df_paris['geo_point_2d']
temp1

0      [48.87689616237872, 2.337460241388529]
1      [48.86790337886785, 2.344107166658533]
2      [48.85941549762748, 2.378741060237548]
3      [48.86305413181178, 2.359361058970589]
4      [48.84896809191946, 2.332670898588416]
5     [48.854228281954754, 2.357361938142205]
6      [48.87252726662346, 2.312582560420059]
7      [48.82871768452136, 2.362468228516128]
8      [48.83515623066034, 2.419807034965275]
9     [48.844508659617546, 2.349859385560182]
10     [48.88686862295828, 2.384694327870042]
11     [48.86318677744551, 2.400819826729021]
12     [48.87602855694339, 2.361112904561707]
13     [48.86039876035177, 2.262099559395783]
14     [48.85608259819694, 2.312438687733857]
15    [48.892735074561706, 2.348711933867703]
16     [48.88733716648682, 2.307485559493426]
17     [48.84015541860987, 2.293559372435076]
18      [48.8626304851685, 2.336293446550539]
19     [48.82899321160942, 2.327100883257538]
Name: geo_point_2d, dtype: object

In [65]:
paris_latlng = df_paris['geo_point_2d'].astype('str')

Spliting the geo_point_2d column into latitude and longitude.

## Latitude

In [66]:
paris_lat = paris_latlng.apply(lambda x: x.split(',')[0])
paris_lat = paris_lat.apply(lambda x: x.lstrip('['))
paris_lat

0      48.87689616237872
1      48.86790337886785
2      48.85941549762748
3      48.86305413181178
4      48.84896809191946
5     48.854228281954754
6      48.87252726662346
7      48.82871768452136
8      48.83515623066034
9     48.844508659617546
10     48.88686862295828
11     48.86318677744551
12     48.87602855694339
13     48.86039876035177
14     48.85608259819694
15    48.892735074561706
16     48.88733716648682
17     48.84015541860987
18      48.8626304851685
19     48.82899321160942
Name: geo_point_2d, dtype: object

## Longitude

In [67]:
paris_lng = paris_latlng.apply(lambda x: x.split(',')[1])
paris_lng = paris_lng.apply(lambda x: x.rstrip(']'))
paris_lng

0      2.337460241388529
1      2.344107166658533
2      2.378741060237548
3      2.359361058970589
4      2.332670898588416
5      2.357361938142205
6      2.312582560420059
7      2.362468228516128
8      2.419807034965275
9      2.349859385560182
10     2.384694327870042
11     2.400819826729021
12     2.361112904561707
13     2.262099559395783
14     2.312438687733857
15     2.348711933867703
16     2.307485559493426
17     2.293559372435076
18     2.336293446550539
19     2.327100883257538
Name: geo_point_2d, dtype: object

In [68]:
paris_geo_lat  = pd.DataFrame(paris_lat.astype(float))
paris_geo_lat.columns=['Latitude']
paris_geo_lng = pd.DataFrame(paris_lng.astype(float))
paris_geo_lng.columns=['Longitude']

In [69]:
paris_combined_data = pd.concat([df_paris.drop('geo_point_2d', axis=1), paris_geo_lat, paris_geo_lng], axis=1)
paris_combined_data

postal_code                  nom_comm nom_dept   Latitude  Longitude
0        75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460
1        75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107
2        75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741
3        75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361
4        75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671
5        75004   PARIS-4E-ARRONDISSEMENT    PARIS  48.854228   2.357362
6        75008   PARIS-8E-ARRONDISSEMENT    PARIS  48.872527   2.312583
7        75013  PARIS-13E-ARRONDISSEMENT    PARIS  48.828718   2.362468
8        75012  PARIS-12E-ARRONDISSEMENT    PARIS  48.835156   2.419807
9        75005   PARIS-5E-ARRONDISSEMENT    PARIS  48.844509   2.349859
10       75019  PARIS-19E-ARRONDISSEMENT    PARIS  48.886869   2.384694
11       75020  PARIS-20E-ARRONDISSEMENT    PARIS  48.863187   2.400820
12       75010  PARIS-10E-ARRONDISSEMENT    PARIS  48.876029   2.361113
13       75016  PARIS-16E-ARRONDISSEMENT    PARIS  48.860399   2.262100
14       75007   PARIS-7E-ARRONDISSEMENT    PARIS  48.856083   2.312439
15       75018  PARIS-18E-ARRONDISSEMENT    PARIS  48.892735   2.348712
16       75017  PARIS-17E-ARRONDISSEMENT    PARIS  48.887337   2.307486
17       75015  PARIS-15E-ARRONDISSEMENT    PARIS  48.840155   2.293559
18       75001  PARIS-1ER-ARRONDISSEMENT    PARIS  48.862630   2.336293
19       75014  PARIS-14E-ARRONDISSEMENT    PARIS  48.828993   2.327101

In [70]:
paris = geocode(address='Paris, France, FR')[0]
paris_lng_coords = paris['location']['x']
paris_lat_coords = paris['location']['y']
print("The geolocation of Paris: ", paris_lat_coords, paris_lng_coords)

The geolocation of Paris:  48.85341000000005 2.3488000000000397


# Visualize the Map of Paris

In [71]:
# Creating the map of Paris
map_Paris= folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)
map_Paris

# adding markers to map
for latitude, longitude, borough, town in zip(paris_combined_data['Latitude'], paris_combined_data['Longitude'], paris_combined_data['nom_comm'], paris_combined_data['nom_dept']):
    label = '{}, {}'.format(town, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='Blue',
        fill=True,
        fill_opacity=0.8
        ).add_to(map_Paris)  
    
map_Paris

## Venues in Paris


Using our previously defined function. Let's get the neaby venues present in each neighbourhood of Paris

In [72]:
venues_in_Paris = getNearbyVenues(paris_combined_data['nom_comm'], paris_combined_data['Latitude'], paris_combined_data['Longitude'])

PARIS-9E-ARRONDISSEMENT
PARIS-2E-ARRONDISSEMENT
PARIS-11E-ARRONDISSEMENT
PARIS-3E-ARRONDISSEMENT
PARIS-6E-ARRONDISSEMENT
PARIS-4E-ARRONDISSEMENT
PARIS-8E-ARRONDISSEMENT
PARIS-13E-ARRONDISSEMENT
PARIS-12E-ARRONDISSEMENT
PARIS-5E-ARRONDISSEMENT
PARIS-19E-ARRONDISSEMENT
PARIS-20E-ARRONDISSEMENT
PARIS-10E-ARRONDISSEMENT
PARIS-16E-ARRONDISSEMENT
PARIS-7E-ARRONDISSEMENT
PARIS-18E-ARRONDISSEMENT
PARIS-17E-ARRONDISSEMENT
PARIS-15E-ARRONDISSEMENT
PARIS-1ER-ARRONDISSEMENT
PARIS-14E-ARRONDISSEMENT


In [73]:
venues_in_Paris.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
1  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
2  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
3  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   
4  PARIS-9E-ARRONDISSEMENT               48.876896                  2.33746   

                    Venue                 Venue Category  
0     Place Saint-Georges                          Plaza  
1              Farine & O                         Bakery  
2                     RAP                   Gourmet Shop  
3  Le Bouclier de Bacchus                       Wine Bar  
4                  So Nat  Vegetarian / Vegan Restaurant

In [74]:
venues_in_Paris.shape

(1272, 5)

We have managed to collect 1324 venue records for the neighbourhoods in Paris


## Grouping by Venue Categories


We need to now see how many Venue Categories are there for further processing

In [75]:
venues_in_Paris.groupby('Venue Category').max()

Neighbourhood  Neighbourhood Latitude  \
Venue Category                                                             
Afghan Restaurant       PARIS-11E-ARRONDISSEMENT               48.859415   
African Restaurant       PARIS-9E-ARRONDISSEMENT               48.876896   
American Restaurant     PARIS-19E-ARRONDISSEMENT               48.892735   
Antique Shop             PARIS-9E-ARRONDISSEMENT               48.876896   
Argentinian Restaurant   PARIS-3E-ARRONDISSEMENT               48.863054   
...                                          ...                     ...   
Wine Bar                 PARIS-9E-ARRONDISSEMENT               48.892735   
Wine Shop                PARIS-3E-ARRONDISSEMENT               48.892735   
Women's Store            PARIS-2E-ARRONDISSEMENT               48.867903   
Zoo                     PARIS-12E-ARRONDISSEMENT               48.835156   
Zoo Exhibit             PARIS-12E-ARRONDISSEMENT               48.835156   

                        Neighbourhood Longitude  \
Venue Category                                    
Afghan Restaurant                      2.378741   
African Restaurant                     2.361113   
American Restaurant                    2.384694   
Antique Shop                           2.337460   
Argentinian Restaurant                 2.359361   
...                                         ...   
Wine Bar                               2.400820   
Wine Shop                              2.361113   
Women's Store                          2.344107   
Zoo                                    2.419807   
Zoo Exhibit                            2.419807   

                                                           Venue  
Venue Category                                                    
Afghan Restaurant                                    Afghanistan  
African Restaurant                             Wally Le Saharien  
American Restaurant                                     Harper's  
Antique Shop                             Hôtel des Ventes Drouot  
Argentinian Restaurant                                     Anahi  
...                                                          ...  
Wine Bar                                               Ô Château  
Wine Shop                                         Trois Fois Vin  
Women's Store                               L'Appartement Sézane  
Zoo                                     Parc zoologique de Paris  
Zoo Exhibit             Grande Serre du Parc Zoologique de Paris  

[205 rows x 4 columns]

198 records in Paris. Paris is also multicultural and diverse as London


# One Hot Encoding


We need to Encode our venue categories to get a better result for our clustering

In [76]:
Paris_venue_cat = pd.get_dummies(venues_in_Paris[['Venue Category']], prefix="", prefix_sep="")
Paris_venue_cat

Afghan Restaurant  African Restaurant  American Restaurant  \
0                     0                   0                    0   
1                     0                   0                    0   
2                     0                   0                    0   
3                     0                   0                    0   
4                     0                   0                    0   
...                 ...                 ...                  ...   
1267                  0                   0                    0   
1268                  0                   0                    0   
1269                  0                   0                    0   
1270                  0                   0                    0   
1271                  0                   0                    0   

      Antique Shop  Argentinian Restaurant  Art Gallery  Art Museum  \
0                0                       0            0           0   
1                0                       0            0           0   
2                0                       0            0           0   
3                0                       0            0           0   
4                0                       0            0           0   
...            ...                     ...          ...         ...   
1267             0                       0            0           0   
1268             0                       0            0           0   
1269             0                       0            0           0   
1270             0                       0            0           0   
1271             0                       0            0           0   

      Arts & Crafts Store  Asian Restaurant  Athletics & Sports  ...  \
0                       0                 0                   0  ...   
1                       0                 0                   0  ...   
2                       0                 0                   0  ...   
3                       0                 0                   0  ...   
4                       0                 0                   0  ...   
...                   ...               ...                 ...  ...   
1267                    0                 0                   0  ...   
1268                    0                 0                   0  ...   
1269                    0                 0                   0  ...   
1270                    0                 0                   0  ...   
1271                    0                 0                   0  ...   

      Udon Restaurant  Vegetarian / Vegan Restaurant  Venezuelan Restaurant  \
0                   0                              0                      0   
1                   0                              0                      0   
2                   0                              0                      0   
3                   0                              0                      0   
4                   0                              1                      0   
...               ...                            ...                    ...   
1267                0                              0                      0   
1268                0                              0                      0   
1269                0                              0                      0   
1270                0                              0                      0   
1271                0                              0                      0   

      Video Game Store  Vietnamese Restaurant  Wine Bar  Wine Shop  \
0                    0                      0         0          0   
1                    0                      0         0          0   
2                    0                      0         0          0   
3                    0                      0         1          0   
4                    0                      0         0          0   
...                ...                    ...       ...        ...   
1267                 0          

In [77]:
Paris_venue_cat['Neighbourhood'] = venues_in_Paris['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [Paris_venue_cat.columns[-1]] + list(Paris_venue_cat.columns[:-1])
Paris_venue_cat = Paris_venue_cat[fixed_columns]

Paris_venue_cat.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-9E-ARRONDISSEMENT                  0                   0   
1  PARIS-9E-ARRONDISSEMENT                  0                   0   
2  PARIS-9E-ARRONDISSEMENT                  0                   0   
3  PARIS-9E-ARRONDISSEMENT                  0                   0   
4  PARIS-9E-ARRONDISSEMENT                  0                   0   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                    0             0                       0            0   
1                    0             0                       0            0   
2                    0             0                       0            0   
3                    0             0                       0            0   
4                    0             0                       0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Udon Restaurant  \
0           0                    0                 0  ...                0   
1           0                    0                 0  ...                0   
2           0                    0                 0  ...                0   
3           0                    0                 0  ...                0   
4           0                    0                 0  ...                0   

   Vegetarian / Vegan Restaurant  Venezuelan Restaurant  Video Game Store  \
0                              0                      0                 0   
1                              0                      0                 0   
2                              0                      0                 0   
3                              0                      0                 0   
4                              1                      0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0                      0         0          0              0    0            0  
1                      0         0          0              0    0            0  
2                      0         0          0              0    0            0  
3                      0         1          0              0    0            0  
4                      0         0          0              0    0            0  

[5 rows x 206 columns]

# Venue categories mean value


We will group the Neighbourhoods and calculate the mean venue categories value in each Neighbourhood

In [78]:
Paris_grouped = Paris_venue_cat.groupby('Neighbourhood').mean().reset_index()
Paris_grouped.head()

Neighbourhood  Afghan Restaurant  African Restaurant  \
0  PARIS-10E-ARRONDISSEMENT            0.00000                0.02   
1  PARIS-11E-ARRONDISSEMENT            0.02439                0.00   
2  PARIS-12E-ARRONDISSEMENT            0.00000                0.00   
3  PARIS-13E-ARRONDISSEMENT            0.00000                0.00   
4  PARIS-14E-ARRONDISSEMENT            0.00000                0.00   

   American Restaurant  Antique Shop  Argentinian Restaurant  Art Gallery  \
0                  0.0           0.0                     0.0          0.0   
1                  0.0           0.0                     0.0          0.0   
2                  0.0           0.0                     0.0          0.0   
3                  0.0           0.0                     0.0          0.0   
4                  0.0           0.0                     0.0          0.0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  ...  Udon Restaurant  \
0     0.00000                  0.0          0.030000  ...              0.0   
1     0.02439                  0.0          0.048780  ...              0.0   
2     0.00000                  0.0          0.000000  ...              0.0   
3     0.00000                  0.0          0.222222  ...              0.0   
4     0.00000                  0.0          0.000000  ...              0.0   

   Vegetarian / Vegan Restaurant  Venezuelan Restaurant  Video Game Store  \
0                           0.01                    0.0           0.00000   
1                           0.00                    0.0           0.02439   
2                           0.00                    0.0           0.00000   
3                           0.00                    0.0           0.00000   
4                           0.00                    0.0           0.00000   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Zoo  Zoo Exhibit  
0               0.000000   0.02000       0.02            0.0  0.0          0.0  
1               0.048780   0.02439       0.00            0.0  0.0          0.0  
2               0.000000   0.00000       0.00            0.0  0.2          0.2  
3               0.206349   0.00000       0.00            0.0  0.0          0.0  
4               0.000000   0.00000       0.00            0.0  0.0          0.0  

[5 rows x 206 columns]

## Top venue categories


Reusing our previously defined function to get the top venue categories in the neighbourhoods of Paris.

In [79]:
# create a new dataframe for Paris
neighborhoods_venues_sorted_paris = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted_paris['Neighbourhood'] = Paris_grouped['Neighbourhood']

for ind in np.arange(Paris_grouped.shape[0]):
    neighborhoods_venues_sorted_paris.iloc[ind, 1:] = return_most_common_venues(Paris_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted_paris.head()

Neighbourhood 1st Most Common Venue  2nd Most Common Venue  \
0  PARIS-10E-ARRONDISSEMENT     French Restaurant                 Bistro   
1  PARIS-11E-ARRONDISSEMENT            Restaurant       Asian Restaurant   
2  PARIS-12E-ARRONDISSEMENT           Zoo Exhibit            Supermarket   
3  PARIS-13E-ARRONDISSEMENT      Asian Restaurant  Vietnamese Restaurant   
4  PARIS-14E-ARRONDISSEMENT     French Restaurant                  Hotel   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0           Coffee Shop                 Hotel                   Café   
1    Italian Restaurant                  Café  Vietnamese Restaurant   
2                Bistro   Monument / Landmark                    Zoo   
3       Thai Restaurant    Chinese Restaurant      French Restaurant   
4                 Plaza             Brasserie     Italian Restaurant   

  6th Most Common Venue 7th Most Common Venue   8th Most Common Venue  \
0           Pizza Place   Japanese Restaurant      Italian Restaurant   
1     French Restaurant                Bakery                  Bistro   
2       Cultural Center          Cupcake Shop              Food Court   
3             Juice Bar              Bus Stop  Furniture / Home Store   
4   Japanese Restaurant           Pizza Place                Tea Room   

  9th Most Common Venue 10th Most Common Venue  
0      Asian Restaurant        Thai Restaurant  
1           Pastry Shop  Performing Arts Venue  
2     Food & Drink Shop               Creperie  
3        Farmers Market   Cambodian Restaurant  
4                  Café   Fast Food Restaurant

# Model Building
## K Means


Let's cluster the city of Paris to roughly 5 to make it easier to analyze.
We use the K Means clustering technique to do so.

In [80]:
# set number of clusters
k_num_clusters = 5

Paris_grouped_clustering = Paris_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans_Paris = KMeans(n_clusters=k_num_clusters, random_state=0).fit(Paris_grouped_clustering)
kmeans_Paris

KMeans(n_clusters=5, random_state=0)

### Labelling Clustered Data

In [81]:
kmeans_Paris.labels_

array([0, 0, 2, 3, 1, 0, 4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0],
      dtype=int32)

In [82]:
neighborhoods_venues_sorted_paris.insert(0, 'Cluster Labels', kmeans_Paris.labels_ +1)

Join paris_combined_data with our neighbourhood venues sorted to add latitude & longitude for each of the neighborhood to prepare it for plotting

In [83]:
paris_data = paris_combined_data

paris_data = paris_data.join(neighborhoods_venues_sorted_paris.set_index('Neighbourhood'), on='nom_comm')

paris_data.head()

postal_code                  nom_comm nom_dept   Latitude  Longitude  \
0       75009   PARIS-9E-ARRONDISSEMENT    PARIS  48.876896   2.337460   
1       75002   PARIS-2E-ARRONDISSEMENT    PARIS  48.867903   2.344107   
2       75011  PARIS-11E-ARRONDISSEMENT    PARIS  48.859415   2.378741   
3       75003   PARIS-3E-ARRONDISSEMENT    PARIS  48.863054   2.359361   
4       75006   PARIS-6E-ARRONDISSEMENT    PARIS  48.848968   2.332671   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0               1     French Restaurant                 Hotel   
1               1     French Restaurant          Cocktail Bar   
2               1            Restaurant      Asian Restaurant   
3               1     French Restaurant           Art Gallery   
4               1                Bakery        Chocolate Shop   

  3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
0   Japanese Restaurant                Bistro                 Bakery   
1              Wine Bar                Bakery               Creperie   
2    Italian Restaurant                  Café  Vietnamese Restaurant   
3           Coffee Shop          Gourmet Shop           Burger Joint   
4              Wine Bar            Restaurant               Fountain   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0              Wine Bar                Lounge          Cocktail Bar   
1                 Hotel            Restaurant       Thai Restaurant   
2     French Restaurant                Bakery                Bistro   
3                Bakery          Cocktail Bar        Sandwich Place   
4           Pastry Shop          Cocktail Bar                Bistro   

  9th Most Common Venue 10th Most Common Venue  
0                 Plaza               Tea Room  
1      Pedestrian Plaza            Coffee Shop  
2           Pastry Shop  Performing Arts Venue  
3   Japanese Restaurant     Italian Restaurant  
4     Electronics Store    Lebanese Restaurant

In [84]:
paris_data_nonan = paris_data.dropna(subset=['Cluster Labels'])

## Visualizing the clustered neighbourhood


Let's plot the clusters

In [85]:
map_clusters_paris = folium.Map(location=[paris_lat_coords, paris_lng_coords], zoom_start=12)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(paris_data_nonan['Latitude'], paris_data_nonan['Longitude'], paris_data_nonan['nom_comm'], paris_data_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + ' ' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.8
        ).add_to(map_clusters_paris)
        
map_clusters_paris

# Examining our Clusters

Clusters 1, 3 and 5

### Cluster 1

In [86]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 1, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
0    PARIS-9E-ARRONDISSEMENT               1     French Restaurant   
1    PARIS-2E-ARRONDISSEMENT               1     French Restaurant   
2   PARIS-11E-ARRONDISSEMENT               1            Restaurant   
3    PARIS-3E-ARRONDISSEMENT               1     French Restaurant   
4    PARIS-6E-ARRONDISSEMENT               1                Bakery   
5    PARIS-4E-ARRONDISSEMENT               1     French Restaurant   
9    PARIS-5E-ARRONDISSEMENT               1     French Restaurant   
10  PARIS-19E-ARRONDISSEMENT               1     French Restaurant   
11  PARIS-20E-ARRONDISSEMENT               1                Bakery   
12  PARIS-10E-ARRONDISSEMENT               1     French Restaurant   
15  PARIS-18E-ARRONDISSEMENT               1     French Restaurant   
17  PARIS-15E-ARRONDISSEMENT               1                 Hotel   
18  PARIS-1ER-ARRONDISSEMENT               1     French Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Hotel   Japanese Restaurant                Bistro   
1           Cocktail Bar              Wine Bar                Bakery   
2       Asian Restaurant    Italian Restaurant                  Café   
3            Art Gallery           Coffee Shop          Gourmet Shop   
4         Chocolate Shop              Wine Bar            Restaurant   
5                  Hotel        Ice Cream Shop        Clothing Store   
9                  Hotel                 Plaza                Bakery   
10                   Bar    Seafood Restaurant               Brewery   
11                 Plaza     French Restaurant                Bistro   
12                Bistro           Coffee Shop                 Hotel   
15                   Bar    Italian Restaurant                 Plaza   
17    Italian Restaurant     French Restaurant   Japanese Restaurant   
18   Japanese Restaurant                 Hotel                 Plaza   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Bakery              Wine Bar                Lounge   
1                Creperie                 Hotel            Restaurant   
2   Vietnamese Restaurant     French Restaurant                Bakery   
3            Burger Joint                Bakery          Cocktail Bar   
4                Fountain           Pastry Shop          Cocktail Bar   
5                    Park           Pastry Shop              Wine Bar   
9      Italian Restaurant           Coffee Shop                   Pub   
10                 Bistro           Supermarket              Beer Bar   
11                  Hotel                   Bar   Japanese Restaurant   
12                   Café           Pizza Place   Japanese Restaurant   
15            Pizza Place                Bistro                 Hotel   
17                 Bakery            Restaurant   Lebanese Restaurant   
18                   Café            Art Museum    Italian Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0           Cocktail Bar                 Plaza               Tea Room  
1        Thai Restaurant      Pedestrian Plaza            Coffee Shop  
2                 Bistro           Pastry Shop  Performing Arts Venue  
3         Sandwich Place   Japanese Restaurant     Italian Restaurant  
4                 Bistro     Electronics Store    Lebanese Restaurant  
5       Pedestrian Plaza                Bistro       Tapas Restaurant  
9                   Café        Science Museum                    Bar  
10                 Hotel  Gym / Fitness Center       Sushi Restaurant  
11    Italian Restaurant                  Café            Pizza Place  
12    Italian Restaurant      Asian Restaurant        Thai Restaurant  
15                  Café            Restaurant      Convenience Store  
17             Brasserie   Arts & Crafts Store      Indian Restaurant  
18        Clothing Store       Thai Restaurant            Coffee Shop

### Cluster 3

In [87]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 3, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
8  PARIS-12E-ARRONDISSEMENT               3           Zoo Exhibit   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8           Supermarket                Bistro   Monument / Landmark   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8                   Zoo       Cultural Center          Cupcake Shop   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8            Food Court     Food & Drink Shop               Creperie

### Cluster 5

In [88]:
paris_data_nonan.loc[paris_data_nonan['Cluster Labels'] == 5, paris_data_nonan.columns[[1] + list(range(5, paris_data_nonan.shape[1]))]]

nom_comm  Cluster Labels 1st Most Common Venue  \
13  PARIS-16E-ARRONDISSEMENT               5                  Lake   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13                  Pool            Art Museum                  Park   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13              Bus Stop           Bus Station         Boat or Ferry   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13     French Restaurant      Recording Studio                  Plaza

# Results and Discussion



The neighbourhoods of London are very mulitcultural. There are a lot of different cusines including Indian, Italian, Turkish and Chinese. London seems to take a step further in this direction by having a lot of Restaurants, bars, juice bars, coffee shops, Fish and Chips shop and Breakfast spots. It has a lot of shopping options too with that of the Flea markets, flower shops, fish markets, Fishing stores, clothing stores. The main modes of transport seem to be Buses and trains. For leisure, the neighbourhoods are set up to have lots of parks, golf courses, zoo, gyms and Historic sites.
Overall, the city of London offers a multicultural, diverse and certainly an entertaining experience.
Paris is relatively small in size geographically. It has a wide variety of cusines and eateries including French, Thai, Cambodian, Asian, Chinese etc. There are a lot of hangout spots including many Restaurants and Bars. Paris has a lot of Bistro's. Different means of public transport in Paris which includes buses, bikes, boats or ferries. For leisure and sight seeing, there are a lot of Plazas, Trails, Parks, Historic sites, clothing shops, Art galleries and Museums. Overall, Paris seems like the relaxing vacation spot with a mix of lakes, historic spots and a wide variety of cusines to try out.



# Conclusion



The purpose of this project was to explore the cities of London and Paris and see how attractive it is to potential tourists and migrants. We explored both the cities based on their postal codes and then extrapolated the common venues present in each of the neighbourhoods finally concluding with clustering similar neighbourhoods together.
We could see that each of the neighbourhoods in both the cities have a wide variety of experiences to offer which is unique in it's own way. The cultural diversity is quite evident which also gives the feeling of a sense of inclusion.
Both Paris and London seem to offer a vacation stay or a romantic gateaway with a lot of places to explore, beautiful landscapes and a wide variety of culture.Overall, it's upto the stakeholders to decide which experience they would prefer more and which would more to their liking.